include/hwmalloc/detail/pool.hpp

```diff
@@ -48,20 +50,25 @@ class pool
     }
 
   private:
-    Context*    m_context;
-    std::size_t m_block_size;
-    std::size_t m_segment_size;
-    std::size_t m_numa_node;
-    bool        m_never_free;
-    std::size_t m_num_reserve_segments;
-    stack_type  m_free_stack;
-    segment_map m_segments;
-    std::mutex  m_mutex;
-    int         m_device_id = 0;
-    bool        m_allocate_on_device = false;
+    Context*                     m_context;
+    std::size_t                  m_block_size;
+    std::size_t                  m_segment_size;
+    std::size_t                  m_numa_node;
+    bool                         m_never_free;
+    std::size_t                  m_num_reserve_segments;
+    stack_type                   m_free_stack;
+    segment_map                  m_segments;
+    std::mutex                   m_mutex;
+    int                          m_device_id = 0;
+    bool                         m_allocate_on_device = false;
+    // Callback executed when a segment is allocated. Can be used to track the number of actual
+    // memory allocations.
+    const segment_alloc_cb_type& m_segment_alloc_cb;
```

Comment:
Please correct me if I'm missing something: I _think_ this will be a dangling reference when using the default value in the `fixed_size_heap` constructor. The `nullptr` will produce a prvalue for the call to the constructor. The `const&` in the constructor will keep the prvalue alive until the call to the constructor finishes. The `const&` member will point to the prvalue that has been destroyed once the constructor call finishes.

Options:
- make the default point to a static variable
- pass and store the callback by value

The second option is seemingly simple, but the reference to the parent class is hidden in the captured `this`. Safe in the default case, but may be too implicit in the non-default case?